File: 2023/python/day2.py
from collections import defaultdict


def get_games():
    with open('../input/input02') as f:
        lines = [line.strip() for line in f.readlines() if len(line) > 0]
        games = {}
        for line in lines:
            game_id = int(line.split(':')[0].replace("Game ", ""))
            sets = [set.strip() for set in line.split(':')[1].split('; ')]
            sets = [set.split(', ') for set in sets]
            game_sets = []
            for set in sets:
                game_set = defaultdict(int, letters=['red', 'green', 'blue'])
                for entry in set:
                    count = int(entry.split(' ')[0])
                    color = entry.split(' ')[1]
                    game_set[color] = count
                game_sets.append(game_set)
            games[game_id] = game_sets
        return games


def is_game_valid(game_sets):
    for game_set in game_sets:
        if game_set['red'] > 12 or game_set['green'] > 13 or game_set['blue'] > 14:
            return False
    return True


def part1():
    print(sum([game_id for game_id, game_sets in get_games().items()
          if is_game_valid(game_sets)]))


def power_game(game_sets):
    red, green, blue = 0, 0, 0
    for game_set in game_sets:
        red = max(red, game_set['red'])
        green = max(green, game_set['green'])
        blue = max(blue, game_set['blue'])
    return red * green * blue


def part2():
    print(sum([power_game(game_sets) for _, game_sets in get_games().items()]))


if __name__ == '__main__':
    part1()
    part2()

File: 2023/deno/day2.ts
// Read and parse input

// const input = await Deno.readTextFile("../input/sample02");
const input = await Deno.readTextFile("../input/input02");
const games = input
  .split("\n")
  .filter((line) => line.length > 0)
  .map((line) => {
    const [game, sets] = line.split(": ");
    const gameId = parseInt(game.replaceAll("Game ", ""));
    const gameSets = sets.split("; ").map((set) => {
      const result: Map<string, number> = new Map();
      set.split(", ").map((cube) => {
        const [count, color] = cube.split(" ");
        result.set(color, parseInt(count));
      });
      return result;
    });
    return { id: gameId, sets: gameSets };
  });

// Part 1: What is the sum of the IDs of those games that are not impossible?

const part1 = games
  .filter(
    (game) =>
      !game.sets.some(
        (set) =>
          (set.get("red") ?? 0) > 12 ||
          (set.get("green") ?? 0) > 13 ||
          (set.get("blue") ?? 0) > 14
      )
  )
  .reduce((acc, game) => acc + game.id, 0);

console.log("Part 1:", part1);

// Part 2: What is the sum of the power of the sets that have the fewest number of cubes of each color

const part2 = games.reduce((acc, game) => {
  let red = 0;
  let green = 0;
  let blue = 0;
  game.sets.forEach((set) => {
    red = Math.max(red, set.get("red") ?? 0);
    green = Math.max(green, set.get("green") ?? 0);
    blue = Math.max(blue, set.get("blue") ?? 0);
  });
  acc += red * green * blue;
  return acc;
}, 0);

console.log("Part 2:", part2);
